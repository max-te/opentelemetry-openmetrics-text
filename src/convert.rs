use std::borrow::Cow;
use std::fmt::{Display, Write};

use opentelemetry::KeyValue;
use opentelemetry_sdk::metrics::data::ResourceMetrics;

pub struct ToOpenMetrics<'a>(pub &'a ResourceMetrics);

impl<'a> ToOpenMetrics<'a> {
    pub const MIME_TYPE: &'static str =
        "application/openmetrics-text; version=1.0.0; charset=utf-8";
}

impl<'a> Display for ToOpenMetrics<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        // let resource_attrs = self.0.resource().into_iter().collect::<Vec<_>>();

        for scope in self.0.scope_metrics() {
            // write_scope_info(f, scope)?;
            let scope_name = scope.scope().name();

            for metric in scope.metrics() {
                let mut name = sanitize_name(metric.name());
                let unit = convert_unit(metric.unit());
                if !unit.is_empty() {
                    name.push('_');
                    name.push_str(&unit);
                    writeln!(f, "# UNIT {name} {unit}")?;
                }
                if !metric.description().is_empty() {
                    writeln!(
                        f,
                        "# HELP {name} {}",
                        escape_label_value(metric.description())
                    )?;
                }
                match metric.data() {
                    opentelemetry_sdk::metrics::data::AggregatedMetrics::F64(metric_data) => {
                        match metric_data {
                            opentelemetry_sdk::metrics::data::MetricData::Gauge(gauge) => {
                                write_gauge(f, name, scope_name, gauge)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Sum(sum) => {
                                write_counter(f, name, scope_name, sum)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Histogram(histogram) => {
                                write_histogram(f, name, scope_name, histogram)?;
                            }
                            // See https://github.com/open-telemetry/opentelemetry-specification/blob/v1.45.0/specification/compatibility/prometheus_and_openmetrics.md#exponential-histograms
                            // for exponential histograms
                            _ => unimplemented!(),
                        }
                    }
                    opentelemetry_sdk::metrics::data::AggregatedMetrics::U64(metric_data) => {
                        match metric_data {
                            opentelemetry_sdk::metrics::data::MetricData::Gauge(gauge) => {
                                write_gauge(f, name, scope_name, gauge)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Sum(sum) => {
                                write_counter(f, name, scope_name, sum)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Histogram(histogram) => {
                                write_histogram(f, name, scope_name, histogram)?;
                            }
                            _ => unimplemented!(),
                        }
                    }
                    opentelemetry_sdk::metrics::data::AggregatedMetrics::I64(metric_data) => {
                        match metric_data {
                            opentelemetry_sdk::metrics::data::MetricData::Gauge(gauge) => {
                                write_gauge(f, name, scope_name, gauge)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Sum(sum) => {
                                write_counter(f, name, scope_name, sum)?;
                            }
                            opentelemetry_sdk::metrics::data::MetricData::Histogram(histogram) => {
                                write_histogram(f, name, scope_name, histogram)?;
                            }
                            _ => unimplemented!(),
                        }
                    }
                }
            }
        }
        Ok(())
    }
}

// fn write_scope_info(
//     f: &mut std::fmt::Formatter<'_>,
//     scope: &opentelemetry_sdk::metrics::data::ScopeMetrics,
// ) -> Result<(), std::fmt::Error> {
//     // https://github.com/open-telemetry/opentelemetry-specification/blob/v1.45.0/specification/compatibility/prometheus_and_openmetrics.md#instrumentation-scope-1
//     let scope_name = sanitize_name(scope.scope().name());
//     let scope_attrs = print_attrs(scope.scope().attributes());
//     writeln!(f, "# TYPE {scope_name} info")?;
//     writeln!(f, "{scope_name}_info{{{}}} 1", scope_attrs)?;
//     Ok(())
// }

fn write_histogram<T: FastDisplay + Copy>(
    f: &mut std::fmt::Formatter<'_>,
    name: String,
    scope_name: &str,
    histogram: &opentelemetry_sdk::metrics::data::Histogram<T>,
) -> Result<(), std::fmt::Error> {
    let scope_name_attrs = &[KeyValue::new("otel_scope_name", scope_name.to_owned())];
    writeln!(f, "# TYPE {name} histogram")?;
    let ts = to_timestamp(histogram.time());
    let created = to_timestamp(histogram.start_time());
    writeln!(
        f,
        "{name}_created{{{attrs}}} {created}",
        attrs = print_attrs(scope_name_attrs.iter())
    )?;
    assert_eq!(
        histogram.temporality(),
        opentelemetry_sdk::metrics::Temporality::Cumulative,
        "Only cumulative Histograms are supported"
    );
    for point in histogram.data_points() {
        let mut attrs = print_attrs(point.attributes().chain(scope_name_attrs.iter()));

        writeln!(
            f,
            "{name}_count{{{attrs}}} {value} {ts}",
            value = point.count().fast_display()
        )?;

        writeln!(
            f,
            "{name}_sum{{{attrs}}} {value} {ts}",
            value = point.sum().fast_display()
        )?;

        // Non-compliant but useful:
        if let Some(min) = point.min() {
            writeln!(
                f,
                "{name}_min{{{attrs}}} {value} {ts}",
                value = min.fast_display()
            )?;
        }
        if let Some(max) = point.max() {
            writeln!(
                f,
                "{name}_max{{{attrs}}} {value} {ts}",
                value = max.fast_display()
            )?;
        }

        if !attrs.is_empty() {
            attrs.push(',');
        }
        for (bound, count) in std::iter::zip(point.bounds(), point.bucket_counts()) {
            writeln!(
                f,
                "{name}_bucket{{{attrs}le=\"{le}\"}} {value} {ts}",
                le = bound.fast_display(),
                value = count.fast_display()
            )?;
        }
        writeln!(
            f,
            "{name}_bucket{{{attrs}le=\"+Inf\"}} {value} {ts}",
            value = point.count().fast_display()
        )?;
    }
    Ok(())
}

fn write_counter<T: FastDisplay + Copy>(
    f: &mut std::fmt::Formatter<'_>,
    name: String,
    scope_name: &str,
    sum: &opentelemetry_sdk::metrics::data::Sum<T>,
) -> Result<(), std::fmt::Error> {
    let scope_name_attrs = &[KeyValue::new("otel_scope_name", scope_name.to_owned())];
    assert_eq!(
        sum.temporality(),
        opentelemetry_sdk::metrics::Temporality::Cumulative,
        "Only cumulative sums are supported"
    );
    if sum.is_monotonic() {
        writeln!(f, "# TYPE {name} counter")?;
        let ts = to_timestamp(sum.time());
        for point in sum.data_points() {
            let attrs = print_attrs(point.attributes().chain(scope_name_attrs));
            writeln!(
                f,
                "{name}_total{{{attrs}}} {value} {ts}",
                value = point.value().fast_display(),
            )?;
        }
        Ok(())
    } else {
        writeln!(f, "# TYPE {name} gauge")?;
        let ts = to_timestamp(sum.time());
        for point in sum.data_points() {
            let attrs = print_attrs(point.attributes().chain(scope_name_attrs));
            writeln!(
                f,
                "{name}{{{attrs}}} {value} {ts}",
                value = point.value().fast_display()
            )?;
        }
        Ok(())
    }
}

fn write_gauge<T: FastDisplay + Copy>(
    f: &mut std::fmt::Formatter<'_>,
    name: String,
    scope_name: &str,
    gauge: &opentelemetry_sdk::metrics::data::Gauge<T>,
) -> Result<(), std::fmt::Error> {
    let scope_name_attrs = &[KeyValue::new("otel_scope_name", scope_name.to_owned())];
    writeln!(f, "# TYPE {name} gauge")?;
    let ts = to_timestamp(gauge.time());
    for point in gauge.data_points() {
        let attrs = print_attrs(point.attributes().chain(scope_name_attrs));
        writeln!(
            f,
            "{name}{{{attrs}}} {value} {ts}",
            value = point.value().fast_display()
        )?;
    }
    Ok(())
}

fn to_timestamp(time: std::time::SystemTime) -> impl Display {
    let ts = time
        .duration_since(std::time::SystemTime::UNIX_EPOCH)
        .expect("Time went backwards")
        .as_secs_f64();
    ts.fast_display()
}

fn print_attrs<'a, I: Iterator<Item = &'a KeyValue>>(attrs: I) -> String {
    let mut result = String::new();
    for attr in attrs {
        if !result.is_empty() {
            result.push(',');
        }
        write!(
            &mut result,
            "{k}=\"{v}\"",
            k = sanitize_name(attr.key.as_str()),
            v = escape_label_value(&attr.value.as_str())
        )
        .unwrap();
    }
    result
}

fn escape_label_value<'a>(value: &'a str) -> Cow<'a, str> {
    let mut bytes = value.as_bytes();
    let first_escape = memchr::memchr3(b'\\', b'"', b'\n', bytes);
    let Some(first_escape) = first_escape else {
        return Cow::Borrowed(value);
    };
    let mut out: Vec<u8> = Vec::with_capacity(bytes.len());
    let (head, tail) = bytes.split_at(first_escape);
    out.extend_from_slice(head);
    out.push(b'\\');
    bytes = tail;

    while let Some(next_escape) = memchr::memchr3(b'\\', b'"', b'\n', bytes) {
        out.extend_from_slice(&bytes[..next_escape]);
        out.push(b'\\');
        bytes = &bytes[next_escape + 1..];
    }
    out.extend_from_slice(bytes);

    Cow::Owned(unsafe {
        // SAFETY: The bytes are valid UTF-8 because they were obtained from a string.
        // Only valid UTF-8 characters are inserted in valid positions.
        String::from_utf8_unchecked(out)
    })
}

fn sanitize_name(name: &str) -> String {
    name.chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || c == ':' {
                c
            } else {
                '_'
            }
        })
        .collect()
}

fn convert_unit(short_unit: &str) -> String {
    match short_unit {
        "" => String::new(),
        "s" => "seconds".to_owned(),
        "ms" => "milliseconds".to_owned(),
        "us" => "microseconds".to_owned(),
        "ns" => "nanoseconds".to_owned(),
        "1" => "ratio".to_owned(),
        "By" => "bytes".to_owned(),
        _ => sanitize_name(short_unit),
    }
}

trait FastDisplay {
    fn fast_display(&self) -> impl Display + Copy + use<Self>;
}

#[derive(Copy, Clone)]
struct RyuDisplay<N: ryu::Float>(N);

impl<N: ryu::Float> Display for RyuDisplay<N> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut buffer = ryu::Buffer::new();
        let formatted = buffer.format(self.0);
        f.write_str(formatted)
    }
}

impl FastDisplay for f64 {
    #[inline]
    fn fast_display(&self) -> impl Display + Copy + use<> {
        RyuDisplay(*self)
    }
}

#[derive(Copy, Clone)]
struct ItoaDisplay<N: itoa::Integer>(N);

impl<N: itoa::Integer> Display for ItoaDisplay<N> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut buffer = itoa::Buffer::new();
        let formatted = buffer.format(self.0);
        f.write_str(formatted)
    }
}

impl FastDisplay for u64 {
    #[inline]
    fn fast_display(&self) -> impl Display + Copy + use<> {
        ItoaDisplay(*self)
    }
}

impl FastDisplay for i64 {
    #[inline]
    fn fast_display(&self) -> impl Display + Copy + use<> {
        ItoaDisplay(*self)
    }
}
