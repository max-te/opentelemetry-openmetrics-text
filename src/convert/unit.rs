/*!
 * OTEL style short unit to Prometheus style long unit conversion
 *
 * Extracted from the opentelemetry-prometheus crate (https://github.com/open-telemetry/opentelemetry-rust/blob/eac368a7e4addbee3b68c27a0eafae59928ad4c7/opentelemetry-prometheus/src/utils.rs)
 * with modifications.
 * Licensed under the Apache-2.0 License, Original Copyright 2025 The opentelemetry-rust Authors
 */

use std::borrow::Cow;

const NON_APPLICABLE_ON_PER_UNIT: [&str; 8] = ["1", "d", "h", "min", "s", "ms", "us", "ns"];

pub(crate) fn get_unit_suffixes(unit: &str) -> Option<Cow<'static, str>> {
    // no unit return early
    if unit.is_empty() {
        return None;
    }

    // direct match with known units
    if let Some(matched) = get_prom_units(unit) {
        return Some(Cow::Borrowed(matched));
    }

    // converting foo/bar to foo_per_bar
    // split the string by the first '/'
    // if the first part is empty, we just return the second part if it's a match with known per unit
    // e.g
    // "test/y" => "per_year"
    // "km/s" => "kilometers_per_second"
    if let Some((first, second)) = unit.split_once('/') {
        return match (
            NON_APPLICABLE_ON_PER_UNIT.contains(&first),
            get_prom_units(first),
            get_prom_per_unit(second),
        ) {
            (true, _, Some(second_part)) | (false, None, Some(second_part)) => {
                Some(Cow::Owned(format!("per_{second_part}")))
            }
            (false, Some(first_part), Some(second_part)) => {
                Some(Cow::Owned(format!("{first_part}_per_{second_part}")))
            }
            _ => None,
        };
    }

    // Unmatched units and annotations are ignored
    // e.g. "{request}"
    None
}

fn get_prom_units(unit: &str) -> Option<&'static str> {
    match unit {
        // Time
        "d" => Some("days"),
        "h" => Some("hours"),
        "min" => Some("minutes"),
        "s" => Some("seconds"),
        "ms" => Some("milliseconds"),
        "us" => Some("microseconds"),
        "ns" => Some("nanoseconds"),

        // Bytes
        "By" => Some("bytes"),
        "KiBy" => Some("kibibytes"),
        "MiBy" => Some("mebibytes"),
        "GiBy" => Some("gibibytes"),
        "TiBy" => Some("tibibytes"),
        "KBy" => Some("kilobytes"),
        "MBy" => Some("megabytes"),
        "GBy" => Some("gigabytes"),
        "TBy" => Some("terabytes"),
        "B" => Some("bytes"),
        "KB" => Some("kilobytes"),
        "MB" => Some("megabytes"),
        "GB" => Some("gigabytes"),
        "TB" => Some("terabytes"),

        // SI
        "m" => Some("meters"),
        "V" => Some("volts"),
        "A" => Some("amperes"),
        "J" => Some("joules"),
        "W" => Some("watts"),
        "g" => Some("grams"),

        // Misc
        "Cel" => Some("celsius"),
        "Hz" => Some("hertz"),
        "1" => Some("ratio"),
        "%" => Some("percent"),
        _ => None,
    }
}

fn get_prom_per_unit(unit: &str) -> Option<&'static str> {
    match unit {
        "s" => Some("second"),
        "m" => Some("minute"),
        "h" => Some("hour"),
        "d" => Some("day"),
        "w" => Some("week"),
        "mo" => Some("month"),
        "y" => Some("year"),
        _ => None,
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn test_get_unit_suffixes() {
        let test_cases = vec![
            // Direct match
            ("g", Some(Cow::Borrowed("grams"))),
            // Per unit
            ("test/y", Some(Cow::Owned("per_year".to_owned()))),
            ("1/y", Some(Cow::Owned("per_year".to_owned()))),
            ("m/s", Some(Cow::Owned("meters_per_second".to_owned()))),
            // No match
            ("invalid", None),
            ("invalid/invalid", None),
            ("seconds", None),
            ("", None),
            // annotations
            ("{request}", None),
        ];
        for (unit, expected_suffix) in test_cases {
            assert_eq!(get_unit_suffixes(unit), expected_suffix);
        }
    }
}
