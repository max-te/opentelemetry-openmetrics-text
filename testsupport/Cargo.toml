[package]
name = "testsupport"
edition = "2024"

[dependencies]
opentelemetry.workspace = true
opentelemetry_sdk = { workspace = true, features = [
    "experimental_metrics_custom_reader",
] }
