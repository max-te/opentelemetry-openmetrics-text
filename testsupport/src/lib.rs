pub mod metric_data;
pub mod reader;
pub mod resource_metrics;
pub mod timestamps;
