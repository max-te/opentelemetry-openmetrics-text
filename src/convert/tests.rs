use std::fmt::Write;
use std::time::UNIX_EPOCH;

use insta::assert_snapshot;
use opentelemetry::KeyValue;
use opentelemetry_sdk::metrics::data::ScopeMetrics;
use testsupport::metric_data::{make_f64_gauge_metric, make_u64_counter_metric, make_f64_histogram_metric};
use testsupport::resource_metrics::make_test_metrics;

use super::*;

#[test]
fn test_write_sanitized_name() {
    let mut output = String::new();

    // Test with valid name
    write_sanitized_name(&mut output, "valid_metric_name").unwrap();
    assert_eq!(output, "valid_metric_name");

    // Test with name containing invalid characters
    output.clear();
    write_sanitized_name(&mut output, "invalid._√§.metric-name").unwrap();
    assert_eq!(output, "invalid_metric_name");

    // Test with name starting with digit
    output.clear();
    write_sanitized_name(&mut output, "1.metric").unwrap();
    assert_eq!(output, "_1_metric");
}

#[test]
fn test_write_escaped() {
    let mut output = String::new();

    // Test with string containing characters that need escaping
    write_escaped(
        &mut output,
        "Line 1\nLine 2\tTabbed\r\nWindows \"quoted\" \\ BS ‚ù§Ô∏è‚Äçü©π",
    )
    .unwrap();
    assert_eq!(
        output,
        "Line 1\\nLine 2\tTabbed\r\\nWindows \\\"quoted\\\" \\\\ BS ‚ù§Ô∏è‚Äçü©π"
    );

    // Test with string not needing escaping
    output.clear();
    write_escaped(&mut output, "Simple string").unwrap();
    assert_eq!(output, "Simple string");
}

#[test]
fn test_hash_attrs() {
    let attrs = [
        KeyValue::new("key1", "value1"),
        KeyValue::new("key2", "value2"),
    ];

    let hash1 = hash_attrs(attrs.iter());

    // Same attributes should produce same hash, order does not matter
    let attrs2 = [
        KeyValue::new("key2", "value2"),
        KeyValue::new("key1", "value1"),
    ];
    let hash2 = hash_attrs(attrs2.iter());

    assert_eq!(hash1, hash2);

    // Different attributes should produce different hash
    let attrs3 = [
        KeyValue::new("key1", "value1"),
        KeyValue::new("key2", "different"),
    ];
    let hash3 = hash_attrs(attrs3.iter());

    assert_ne!(hash1, hash3);
}

#[test]
fn test_write_attrs() {
    let mut output = String::new();
    let attrs = [
        KeyValue::new("key1", "value1"),
        KeyValue::new("key2", "value2"),
    ];

    write_attrs(&mut output, attrs.iter()).unwrap();
    assert_eq!(output, "key1=\"value1\",key2=\"value2\"");

    // Test with attributes containing characters that need escaping
    output.clear();
    let attrs_with_escapes = [
        KeyValue::new("key1", "value\nwith\nnewlines"),
        KeyValue::new("key2", "value\"with\"quotes"),
    ];

    write_attrs(&mut output, attrs_with_escapes.iter()).unwrap();
    assert_eq!(
        output,
        "key1=\"value\\nwith\\nnewlines\",key2=\"value\\\"with\\\"quotes\""
    );
}

#[test]
fn test_make_scope_name_attrs() {
    let scope_name = "test_scope";
    let attr = make_scope_name_attrs(scope_name);

    if cfg!(feature = "otel_scope_info") {
        assert!(attr.is_some());
        if let Some(kv) = attr {
            assert_eq!(kv.key.as_str(), "otel_scope_name");
            assert_eq!(kv.value.as_str(), "test_scope");
        }
    } else {
        assert!(attr.is_none());
    }
}

#[test]
fn test_to_timestamp() {
    use std::time::{Duration, UNIX_EPOCH};

    // Test with a known timestamp
    let time = UNIX_EPOCH + Duration::from_secs(1625097600);
    let timestamp = to_timestamp(time);
    let mut output = String::new();
    write!(output, "{}", timestamp).unwrap();
    assert_eq!(output, "1625097600");
}

#[cfg(feature = "otel_scope_info")]
#[test]
fn test_write_otel_scope_info() {
    let resource_metrics = make_test_metrics();
    let scopes: Vec<&ScopeMetrics> = resource_metrics.scope_metrics().collect();

    let mut output = String::new();
    write_otel_scope_info(&mut output, &scopes).unwrap();

    assert!(output.contains("# TYPE otel_scope info"));
    assert!(output.contains("otel_scope_info{"));
    assert!(output.contains("otel_scope_name=\"meter.1\""));
}

#[test]
fn test_get_type() {
    let resource_metrics = make_test_metrics();
    let scopes: Vec<&ScopeMetrics> = resource_metrics.scope_metrics().collect();

    for scope in scopes {
        for metric in scope.metrics() {
            let result = get_type(metric.data());
            assert!(result.is_ok());

            // Check that the type is one of the expected values
            let type_str = result.unwrap();
            assert!(
                type_str == "gauge" || type_str == "counter" || type_str == "histogram",
                "Unexpected metric type: {}",
                type_str
            );
        }
    }
}

#[test]
fn test_write_values() {
    let resource_metrics = make_test_metrics();
    let scopes: Vec<&ScopeMetrics> = resource_metrics.scope_metrics().collect();

    for scope in scopes {
        let scope_name = scope.scope().name();

        for metric in scope.metrics() {
            let mut output = String::new();
            let mut temp_buffer = String::from("staledata");
            let result = write_values(
                &mut output,
                &mut temp_buffer,
                scope_name,
                metric.data(),
                metric.name(),
            );

            assert!(result.is_ok());
            assert!(!output.is_empty());
            assert!(output.contains(metric.name()));
        }
    }
}

#[test]
fn test_write_gauge() {
    let metric = make_f64_gauge_metric(vec![
        (4.2, vec![KeyValue::new("kk", "v1")]),
        (4.23, vec![KeyValue::new("kk", "v2")]),
    ]);
    let ts = metric
        .time()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
        .to_string();

    let mut output = String::new();
    let mut temp_buffer = String::from("staledata");

    let result = write_gauge(&mut output, "mygauge", "myscope", &mut temp_buffer, &metric);
    let output = output.replace(&ts, "<TIMESTAMP>");

    assert!(result.is_ok());
    assert_snapshot!(output);
}

#[test]
fn test_write_counter() {
    let metric = make_u64_counter_metric(vec![(125, vec![KeyValue::new("kk", "v1")])]);
    let ts = metric
        .time()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
        .to_string();

    let mut output = String::new();
    let mut temp_buffer = String::from("staledata");

    let result = write_counter(
        &mut output,
        "mycounter",
        "myscope",
        &mut temp_buffer,
        &metric,
    );
    let output = output.replace(&ts, "<TIMESTAMP>");

    assert!(result.is_ok());
    assert_snapshot!(output);
}

#[test]
fn test_write_histogram() {
    let metric = make_f64_histogram_metric(vec![
        (125.0, vec![KeyValue::new("kk", "v1")]),
        (125.0, vec![KeyValue::new("kk", "v2")]),
        (25.0, vec![KeyValue::new("kk", "v1")]),
        (0.0, vec![KeyValue::new("kk", "v1")]),
        (25.0, vec![KeyValue::new("kk", "v2")]),
    ]);
    let ts = metric
        .time()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
        .to_string();
    let start_ts = metric
        .start_time()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
        .to_string();

    let mut output = String::new();
    let mut temp_buffer = String::from("staledata");

    let result = write_histogram(
        &mut output,
        "myhistogram",
        "myscope",
        &mut temp_buffer,
        &metric,
    );
    let output = output.replace(&ts, "<TIMESTAMP>");
    let output = output.replace(&start_ts, "<START_TIMESTAMP>");

    assert!(result.is_ok());
    assert_snapshot!(output);
}
