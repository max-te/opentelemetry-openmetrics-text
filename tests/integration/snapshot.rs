use std::time::SystemTime;

use insta::assert_snapshot;
use opentelemetry::time::now;
use opentelemetry_openmetrics::convert::WriteOpenMetrics;

use crate::testsupport::make_test_metrics;

#[test]
fn matches_snapshot() {
    let metrics = make_test_metrics();
    let time = now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_secs()
        .to_string();
    let formatted = metrics.to_openmetrics_string().unwrap();
    let redacted = formatted
        .lines()
        .map(|line| {
            line.split_whitespace()
                .map(|t| String::from(if t.starts_with(&time) { "NOW.sthg" } else { t }))
                .reduce(|acc, e| format!("{acc} {e}"))
                .unwrap_or_default()
        })
        .collect::<Vec<String>>()
        .join("\n");
    assert_snapshot!(redacted);
}
