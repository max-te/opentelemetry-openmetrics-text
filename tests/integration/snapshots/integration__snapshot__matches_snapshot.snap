---
source: tests/integration/snapshot.rs
expression: redacted
---
# TYPE otel_scope info
otel_scope_info{otel_scope_name="meter.1",otel_scope_version=""} 1
# TYPE f64_gauge gauge
# HELP f64_gauge A \"gauge\"\nFor testing
f64_gauge{kk="v1",otel_scope_name="meter.1"} 4.22 NOW.sthg
f64_gauge{kk="v2",otel_scope_name="meter.1"} 4.23 NOW.sthg
# TYPE histo histogram
histo_created{otel_scope_name="meter.1"} NOW.sthg NOW.sthg
histo_count{otel_scope_name="meter.1"} 4 NOW.sthg
histo_sum{otel_scope_name="meter.1"} 15.7 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="0.0"} 1 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="5.0"} 3 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="10.0"} 3 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="25.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="50.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="75.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="100.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="250.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="500.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="750.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="1000.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="2500.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="5000.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="7500.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="10000.0"} 4 NOW.sthg
histo_bucket{otel_scope_name="meter.1",le="+Inf"} 4 NOW.sthg
# TYPE u64_counter_seconds counter
# UNIT u64_counter_seconds seconds
u64_counter_seconds_total{otel_scope_name="meter.1"} 125 NOW.sthg
# EOF
